---
title: "Let's metabaR-F!"
author: "Boyer F., Benoiston A., Donald J., Lionnet C., Zinger L."
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Let's metabaR-F}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: metabaRF.bib
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

![](../metabaRF.png){width=100%}

#Introduction

`metabaR-F` is an R package which supports the importing, handling and post-bioinformatics evaluation and improvement of metabarcoding data quality. It provides a suite of functions to identify and filter common molecular artifacts produced during the experimental workflow, from sampling to sequencing, ideally using experimental controls. 

Due to its simple structure, `metabaR-F` can easily be used in combination with other R packages commonly used for ecological analysis (`vegan`, `ade4`, `ape`, `picante`, etc.). In addition, it provides flexible graphical systems based on `ggplot2` to vizualize data from both an ecological and experimental perspective.

#Dependencies and Installation

`metabaR-F` relies on basic R functions and data structures so as to maximize fexibility and transposability across other packages. It has a minimal number of dependencies to essential R packages : 

- `ggplot2`, `cowplot`, and `igraph` for vizualization purposes
- `reshape2` for data manipulation purposes   
- `vegan` for basic data analyses <span style="color:mediumseagreen"> LZ:... ade4?</span>
- `seqinr` and `biomformat` for data import or formatting.      

To install `metabaR-F`, use : 

```{r install, eval=FALSE}
install.packages("devtools")
devtools::install_github("metabaRfactory/metabaRffe")
```

And then load the package   
```{r loadpackage}
library(metabaRffe) # modify the name once we'll all agree on that
```

#Package overview

![MetabarF over](../metabaRF_overview.png){width=90%}

<span style="color:mediumseagreen"> LZ: --- could include a section here with a figure showing the different steps of data production and to what the different terms we use (e.g. a pcr, a biological sample, a MOTU, a read, tag, primers, etc.) corresponds </span>   

##Data format and structure

The basic dataformat used in `metabaR-F` is a `metabarlist`, consisting of a list of four tables:   

- `reads`: a `matrix` consisting of PCRs as rows, and molecular operational taxonomic units (MOTUs) as columns. The number of reads for MOTUs are given in each cell, with 0 corresponding to no reads.   

- `motus`: a `data.frame` where MOTUs are listed as rows, and their attributes as columns. Examples of attributes include taxonomic information, but could also include any information collected during bioinformatic analysis. A mandatory field in this table is "sequence", i.e. the DNA sequence representative of the MOTU.    

- `pcrs`: a `data.frame` consisting of PCRs as rows, and PCR attributes as columns. This table is particularly important in `metabaR-F`, as it contains all the information related to the sequencing/pcr protocols and design that are necessary to assess and improve the quality of metabarcoding data [@taberlet2018environmental; @zinger2019dna]. This table can also include information related to the PCR design, such as the tag combinations, the primers used, the well and plate of each PCR, etc. Mandatory fields are:    
    - `sample_id`: a vector indicating the biological sample origin of each PCR
    - `type` : type of pcr between an amplification of a sample or of an experimental control. Only two values allowed: `"sample"` or `"control"`.     
    - `control_type` : type of controls. Mandatory values are listed below and should be attributed as follows:   
        - `NA` if `type="sample"`, i.e. for any pcr amplification of a biological sample.    
        - `"extraction"` for DNA extraction negative controls, i.e. pcr amplification of an extraction where sample was replaced by extraction buffer.   
        - `"pcr"` for PCR negative controls, i.e. pcr amplification where the DNA template was replaced by pcr buffer or sterile water.    
        - `"sequencing"` for sequencing negative controls, i.e. sequencing reads present in the data and that were assigned to unused tag/library index combinations.     
        - `"positive"` for DNA extraction or PCR positive controls, i.e. pcr amplifications of known biological samples or DNA template (e.g. mock community).     

- `samples`: a `data.frame` consisting of biological samples as rows, and associated contextual information as columns. Such information includes e.g. geographic coordinates, abiotic parameters, treatment, etc. This table does not include information on the DNA metabarcoding experimental controls, which can only be found in `pcrs`.


###Function Types 

`metabaR-F` provides a range of function types:

- Import and formating functions to import DNA metabarcoding data from common bioinformatic pipelines (OBITools, <span style="color:mediumseagreen">more to come</span>) or more generally to any data formatted into 4 tables corresponding to the `reads`, `motus`, `pcrs`, `samples` mentionned above.   
- Functions for data curation. These are often absent from most bioinformatic pipelines, as they aim at detecting and allow the tagging potential molecular artifacts such as contaminants or dysfynctional pcrs.    
- Functions for visualizing the data under both ecological (e.g. type of samples) and experimental (e.g. type of controls, distribution across the PCR plate design) perspectives.     
- Functions to manipulate the `metabarlist` object, such as selection of a subset, or data aggregation.   
- <span style="color:mediumseagreen">LZ: Blablabla</span>


###Example dataset

An example data set is provided to show how the package can be used to assess and improve the data quality.  

The `soil_euk` dataset is a `metabarlist`. The data were obtained from an environmental DNA (eDNA) metabarcoding experiment aiming to assess the diversity of soil eukaryotes in French Guiana in two sites corresponding to two contrasting habitats:    
- Mana, a site located in a white sand forest, characterized by highly oligotrophic soils and tree species adapted to the harsh local conditions.    
- Petit Plateau, a site located in the pristine rainforest of the Nouragues natural reserve characterized by *terra firme* soils richer in clay and organic matter. 

![](soil_euk_loc.png){width=50%} ![](soil_euk_plots.png){width=36%}

At each site sample collection were conducted at 16 sampling points separated from one another by 20 m and arranged in a grid across a 1 ha plot. At each sampling point, two types of environmental matrix were sampled: soil and litter. One soil sample corresponds to a composite sample of five soil cores. One litter sample corresponds to ca. 1 m2 of litter collected from the forest floor. A total of 64 DNA extracts (i.e. 16 sampling points x 2 sites x 2 types of environmental matrix, i.e. soil and litter) were thus produced, in addition to four DNA extraction controls (one per site and environmental matrix).
 
For each DNA extract, a short region of the 18S rRNA [using the primer pair Euka02 in @taberlet2018environmental] was amplified by PCR in quadruplicate, following the protocol described in @zinger2019body. The resulting amplicons were pooled and sequenced on an Illumina HiSeq platform, using paired-end technology.
 
The total experiment hence resulted in 384 amplicons as follows:

- 256 pcrs corresponding to biological samples (16 sampling points x 2 sites x 2 environmental matrices x 4 pcr replicates).    
- 16 pcrs corresponding to extraction negative controls (4 extraction negative controls x 4 pcr replicates).    
- 32 pcrs corresponding to PCR negative controls (8 pcr negative controls x 4 pcr replicates).    
- 48 sequencing negative controls (monitoring of 48 unused tag combinations).    
- 32 positive controls obtained from a DNA template composed of a mix of 16 plant species (8 pcr positive controls at different dilutions x 4 pcr replicates).   

 <span style="color:mediumseagreen"> LZ -- We might need to include a description of the PCR plate design here too, not everyone is used to this kind of approach </span> 

The retrieved data were then processed using the OBITools [@boyer:2016:00] and SUMACLUST [@mercier:2013:00] packages. Briefly, paired-end reads were assembled, assigned to their respective samples/marker and dereplicated. Low-quality sequences (containing Ns, shorter than 50 bp or singletons) were excluded. The remaining sequences were clustered into molecular operational taxonomic units (MOTUs) using SUMACLUST at a sequence similarity threshold of 97%. The representative sequence of each motu (i.e. the most abundant sequence) was assigned a taxonomic clade using a database built from the EMBL (release 136) with the ecoPCR program [@Ficetola:2010:00].

<span style="color:mediumseagreen"> LZ -- not sure here that we need to explain everything, this is just a copy paste from the help page, we might need to lighten this document a bit. See also depending on what the pkgdown package offers for this </span>

This description can also be found in the `soil_euk` help page: 

```{r help}
?soil_euk
```


##References

